File: src/analyzer/eslint-compat/if-chain-root-offset.test.ts
import { describe, expect, test } from "vitest";

import type { AstNode } from "../../ir/primitive/ast-node.js";
import { AST_TYPE } from "../../parser/ast-type.js";
import type { PathEntry } from "../walk/path-entry.js";
import { ifChainRootOffset } from "./if-chain-root-offset.js";
import type { NodeLike } from "./node-like.js";

function ifNode(start: number): NodeLike {
  return { type: AST_TYPE.IfStatement, start } as const satisfies NodeLike;
}

function blockNode(start: number): NodeLike {
  return { type: AST_TYPE.BlockStatement, start } as const satisfies NodeLike;
}

function entry(node: NodeLike, key: string | null): PathEntry {
  return { node: node as unknown as AstNode, key };
}

describe("ifChainRootOffset", () => {
  test("returns null when parent is null", () => {
    expect(ifChainRootOffset(null, "consequent", [])).toBeNull();
  });

  test("returns null when parent is not an IfStatement", () => {
    expect(ifChainRootOffset(blockNode(0), "consequent", [])).toBeNull();
  });

  test("returns null for keys other than consequent / alternate", () => {
    const parent = ifNode(10);
    expect(ifChainRootOffset(parent, "test", [])).toBeNull();
  });

  test("returns null for a standalone if (parent at top level)", () => {
    const parent = ifNode(10);
    const program = { type: AST_TYPE.Program, start: 0 } as const;
    const path: readonly PathEntry[] = [
      entry(program as unknown as NodeLike, null),
      entry(parent, "body"),
    ];
    expect(ifChainRootOffset(parent, "consequent", path)).toBeNull();
  });

  test("returns outer offset for inner-if consequent in else-if chain (1 step)", () => {
    const outer = ifNode(10);
    const inner = ifNode(40);
    const program = { type: AST_TYPE.Program, start: 0 } as const;
    const path: readonly PathEntry[] = [
      entry(program as unknown as NodeLike, null),
      entry(outer, "body"),
      entry(inner, "alternate"),
    ];
    expect(ifChainRootOffset(inner, "consequent", path)).toBe(10);
  });

  test("returns outer offset for inner-if alternate in else-if chain (1 step)", () => {
    const outer = ifNode(10);
    const inner = ifNode(40);
    const program = { type: AST_TYPE.Program, start: 0 } as const;
    const path: readonly PathEntry[] = [
      entry(program as unknown as NodeLike, null),
      entry(outer, "body"),
      entry(inner, "alternate"),
    ];
    expect(ifChainRootOffset(inner, "alternate", path)).toBe(10);
  });

  test("walks back through multiple chained alternates to the outermost if (2 steps)", () => {
    const outermost = ifNode(5);
    const middle = ifNode(40);
    const innermost = ifNode(80);
    const program = { type: AST_TYPE.Program, start: 0 } as const;
    const path: readonly PathEntry[] = [
      entry(program as unknown as NodeLike, null),
      entry(outermost, "body"),
      entry(middle, "alternate"),
      entry(innermost, "alternate"),
    ];
    expect(ifChainRootOffset(innermost, "consequent", path)).toBe(5);
  });

  test("does not walk past a non-IfStatement ancestor", () => {
    // outer-if's alternate is a BlockStatement which contains a separate
    // inner-if; that inner-if is NOT a chain continuation.
    const outer = ifNode(10);
    const innerBlock = blockNode(40);
    const inner = ifNode(45);
    const program = { type: AST_TYPE.Program, start: 0 } as const;
    const path: readonly PathEntry[] = [
      entry(program as unknown as NodeLike, null),
      entry(outer, "body"),
      entry(innerBlock, "alternate"),
      entry(inner, "body"),
    ];
    expect(ifChainRootOffset(inner, "consequent", path)).toBeNull();
  });

  test("does not walk when current if sits in the consequent slot of an outer if", () => {
    // An IfStatement nested inside an outer if's *consequent* (rather than
    // alternate) is independent, not a chain continuation.
    const outer = ifNode(10);
    const consBlock = blockNode(20);
    const inner = ifNode(25);
    const program = { type: AST_TYPE.Program, start: 0 } as const;
    const path: readonly PathEntry[] = [
      entry(program as unknown as NodeLike, null),
      entry(outer, "body"),
      entry(consBlock, "consequent"),
      entry(inner, "body"),
    ];
    expect(ifChainRootOffset(inner, "consequent", path)).toBeNull();
  });

  test("falls back to 0 when chainTop has no start offset", () => {
    const outer = { type: AST_TYPE.IfStatement } as const satisfies NodeLike;
    const inner = ifNode(40);
    const path: readonly PathEntry[] = [
      entry(outer, "body"),
      entry(inner, "alternate"),
    ];
    expect(ifChainRootOffset(inner, "consequent", path)).toBe(0);
  });
});

File: crates/unsnarl-analyzer/src/if_chain_root_offset_test.rs
use unsnarl_oxc_parity::AstType;

use crate::testing::{ast_node, ast_node_with_end, entry};

use super::if_chain_root_offset;

#[test]
fn returns_none_when_parent_is_null() {
    assert_eq!(if_chain_root_offset(None, Some("consequent"), &[]), None);
}

#[test]
fn returns_none_when_parent_is_not_an_if_statement() {
    let block = ast_node(AstType::BlockStatement, 0);
    assert_eq!(
        if_chain_root_offset(Some(&block), Some("consequent"), &[]),
        None
    );
}

#[test]
fn returns_none_for_keys_other_than_consequent_or_alternate() {
    let parent = ast_node(AstType::IfStatement, 10);
    assert_eq!(if_chain_root_offset(Some(&parent), Some("test"), &[]), None);
}

#[test]
fn returns_none_for_standalone_if() {
    let parent = ast_node_with_end(AstType::IfStatement, 10, 30);
    let path = vec![
        entry(ast_node(AstType::Program, 0), None),
        entry(parent.clone(), Some("body")),
    ];
    assert_eq!(
        if_chain_root_offset(Some(&parent), Some("consequent"), &path),
        None
    );
}

#[test]
fn returns_outer_offset_for_inner_if_consequent_one_step() {
    let outer = ast_node_with_end(AstType::IfStatement, 10, 100);
    let inner = ast_node_with_end(AstType::IfStatement, 40, 80);
    let path = vec![
        entry(ast_node(AstType::Program, 0), None),
        entry(outer, Some("body")),
        entry(inner.clone(), Some("alternate")),
    ];
    assert_eq!(
        if_chain_root_offset(Some(&inner), Some("consequent"), &path),
        Some(10)
    );
}

#[test]
fn returns_outer_offset_for_inner_if_alternate_one_step() {
    let outer = ast_node_with_end(AstType::IfStatement, 10, 100);
    let inner = ast_node_with_end(AstType::IfStatement, 40, 80);
    let path = vec![
        entry(ast_node(AstType::Program, 0), None),
        entry(outer, Some("body")),
        entry(inner.clone(), Some("alternate")),
    ];
    assert_eq!(
        if_chain_root_offset(Some(&inner), Some("alternate"), &path),
        Some(10)
    );
}

#[test]
fn walks_back_through_multiple_chained_alternates() {
    let outermost = ast_node_with_end(AstType::IfStatement, 5, 200);
    let middle = ast_node_with_end(AstType::IfStatement, 40, 180);
    let innermost = ast_node_with_end(AstType::IfStatement, 80, 150);
    let path = vec![
        entry(ast_node(AstType::Program, 0), None),
        entry(outermost, Some("body")),
        entry(middle, Some("alternate")),
        entry(innermost.clone(), Some("alternate")),
    ];
    assert_eq!(
        if_chain_root_offset(Some(&innermost), Some("consequent"), &path),
        Some(5)
    );
}

#[test]
fn does_not_walk_past_a_non_if_statement_ancestor() {
    let outer = ast_node_with_end(AstType::IfStatement, 10, 200);
    let inner_block = ast_node_with_end(AstType::BlockStatement, 40, 100);
    let inner = ast_node_with_end(AstType::IfStatement, 45, 90);
    let path = vec![
        entry(ast_node(AstType::Program, 0), None),
        entry(outer, Some("body")),
        entry(inner_block, Some("alternate")),
        entry(inner.clone(), Some("body")),
    ];
    assert_eq!(
        if_chain_root_offset(Some(&inner), Some("consequent"), &path),
        None
    );
}

#[test]
fn does_not_walk_when_current_if_sits_in_consequent_slot_of_outer() {
    let outer = ast_node_with_end(AstType::IfStatement, 10, 200);
    let cons_block = ast_node_with_end(AstType::BlockStatement, 20, 90);
    let inner = ast_node_with_end(AstType::IfStatement, 25, 70);
    let path = vec![
        entry(ast_node(AstType::Program, 0), None),
        entry(outer, Some("body")),
        entry(cons_block, Some("consequent")),
        entry(inner.clone(), Some("body")),
    ];
    assert_eq!(
        if_chain_root_offset(Some(&inner), Some("consequent"), &path),
        None
    );
}

#[test]
fn falls_back_to_zero_when_chain_top_has_zero_start() {
    let outer = ast_node_with_end(AstType::IfStatement, 0, 100);
    let inner = ast_node_with_end(AstType::IfStatement, 40, 80);
    let path = vec![
        entry(outer, Some("body")),
        entry(inner.clone(), Some("alternate")),
    ];
    assert_eq!(
        if_chain_root_offset(Some(&inner), Some("consequent"), &path),
        Some(0)
    );
}
